In [ ]:
#Problem Statement -
#Build a machine learning model to detect image digit from mnist dataset
#a. Try logistic regression, decision tree, and svm and see which algorithm is performing
#better.
#b. Infer your observations and insights and prepare a report. Make sure it has a
#comparison table at the end.
#Sample mnist dataset
#5 0 4 1 9 2 1 3 1 4

In [ ]:
from PIL import Image
import numpy as np
from google.colab import drive
drive.mount('/content/drive')


def load_mnist_images(filename):
    with open(filename,'rb') as f:
        data = np.frombuffer(f.read(), np.uint8, offset=16)
    return data.reshape(-1, 28, 28)

def load_mnist_labels(filename):
    with open(filename, 'rb') as f:
        data = np.frombuffer(f.read(), np.uint8, offset=8)
    return data

# Load training images and labels
x_train = load_mnist_images("/train-images.idx3-ubyte")
y_train = load_mnist_labels("/train-labels.idx1-ubyte")

# Load testing images and labels
x_test = load_mnist_images("/t10k-images.idx3-ubyte")
y_test = load_mnist_labels("/t10k-labels.idx1-ubyte")

# Flatten and normalize images
x_train = x_train.reshape(-1, 28*28) / 255.0
x_test = x_test.reshape(-1, 28*28) / 255.0


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
logistic_model=LogisticRegression(max_iter=1000)
decision_model=DecisionTreeClassifier()
svm_model=SVC(kernel='linear',C=1.0,random_state=42)
#train the models
logistic_model.fit(x_train,y_train)
decision_model.fit(x_train,y_train)
svm_model.fit(x_train,y_train)
#Do the predictions
y_predict1=logistic_model.predict(x_test)
y_predict2=decision_model.predict(x_test)
y_predict3=svm_model.predict(x_test)
#calculate the accuracy
accuracy1=accuracy_score(y_test,y_predict1)
accuracy2=accuracy_score(y_test,y_predict2)
accuracy3=accuracy_score(y_test,y_predict3)
#print the accuracy
print(f"Accuracy {accuracy1*100:.2f}%")
print(f"Accuracy {accuracy2*100:.2f}%")
print(f"Accuracy {accuracy3*100:.2f}%")



Accuracy 92.56%
Accuracy 87.98%
Accuracy 94.04%



MY OBERSERVATIONS AND INSIGHTS ABOUT THIS PROGECT



By performing logistic_model,decision_model and svm_model,I obeserved
that svm_model is more better for mnist dataset program because I calculate Accuracy for these models,so that basing on the Accuracy
I found the better method.
1. Accuracy of svm_model is 94.04%
2.Accuracy of logistic_model is 92.56%
3.Accuracy of decision_model is 87.98%

so,svm_model algorithm is better to use to detect a image from minist dataset


Comparision Table


S.NO

MODEL

ACCURACY

1.

Logistic_model

92.56%

2.

Decision_model

87.98%

3.

Svm_model

94.04%









